### ISSUE TITLE
Can't get solaar to fn-swap on boot

### ISSUE THREAD

**Issue by webodan** [pwr-Solaar/Solaar#1782]
**Information**

Basically title. The command itself, solaar config "K400 Plus" fn-swap false) does swap the fn keys correctly, but no matter how i try to automate it on my boot running XFCE, it won't swap it (adding it both /etc/X11/xinit/xtinirc and the session startup menu appear to run the program, but it doesn't actually swap the fn keys). I would like to not have to run the program manually so it changes the fn keys by itself, if possible, but I was simply unable to.

I have it set to false in ~/.config/solaar/config.json and config.yaml as well.

- Solaar version (`solaar --version` or `git describe --tags` if cloned from this repository): in both 1.1.0 from Mint and 1.1.5 from stable ppa
- Distribution: mint 21
- Kernel version (ex. `uname -srmo`): Linux 5.19.11-tkg-bmq
- Output of `solaar show`:
Solaar version 1.1.5

<details>

```

Receptor Unifying
  Device path  : /dev/hidraw8
  USB id       : 046d:C52B
  Serial       : 42F4925D
    Firmware   : 24.11.B0036
    Bootloader : 02.09
    Other      : AA.AC
  Has 1 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=195

  1: Wireless Touch Keyboard K400 Plus
     Device path  : /dev/hidraw9
     WPID         : 404D
     Codename     : K400 Plus
     Kind         : keyboard
     Protocol     : HID++ 4.1
     Polling rate : 8 ms (125Hz)
     Serial number: 304E66B3
     Model ID:      404D00000000
     Unit ID:       0F18F9DB
        Bootloader: BOT 22.02.B0002
          Firmware: RQK 63.02.B0016
             Other: 
     The power switch is located on the top edge.
     Supports 23 HID++ 2.0 features:
         0: ROOT                   {0000} V0     
         1: FEATURE SET            {0001} V0     
         2: DEVICE FW VERSION      {0003} V1     
            Firmware: Bootloader BOT 22.02.B0002 404D6217AD40
            Firmware: Firmware RQK 63.02.B0016 404D6217AD40
            Firmware: Other   
            Unit ID: 0F18F9DB  Model ID: 404D00000000  Transport IDs: {'wpid': '404D'}
         3: DEVICE NAME            {0005} V0     
            Name: Wireless Touch Keyboard K400 Plus
            Kind: keyboard
         4: WIRELESS DEVICE STATUS {1D4B} V0     
         5: RESET                  {0020} V0     
         6: BATTERY STATUS         {1000} V0     
            Battery: 90%, discharging, next level 50%.
         7: REPROG CONTROLS V4     {1B04} V0     
            Desvio de Tecla/Botón (saved): {Volume Up:Normal, Volume Down:Normal, Mute:Normal, Play/Pause:Normal, Next:Normal, Previous:Normal, Application Switcher:Normal, My Home:Normal, Media Player:Normal, Search:Normal, Sleep:Normal, Contextual Menu:Normal, Back:Normal, Show Desktop:Normal, Maximize Window:Normal, Switch Screen:Normal, Fn Left Click:Normal, Second Left Click:Normal}
            Desvio de Tecla/Botón        : {Volume Up:Normal, Volume Down:Normal, Mute:Normal, Play/Pause:Normal, Next:Normal, Previous:Normal, Application Switcher:Normal, My Home:Normal, Media Player:Normal, Search:Normal, Sleep:Normal, Contextual Menu:Normal, Back:Normal, Show Desktop:Normal, Maximize Window:Normal, Switch Screen:Normal, Fn Left Click:Normal, Second Left Click:Normal}
         8: SWAP BUTTON CANCEL     {2005} V0     
         9: NEW FN INVERSION       {40A2} V0     
            Fn-swap: disabled
            Fn-swap default: enabled
            Intercambiar función Fx (saved): False
            Intercambiar función Fx        : False
        10: ENCRYPTION             {4100} V0     
        11: KEYBOARD DISABLE KEYS  {4521} V0     
            Desactivar teclas (saved): {Caps Lock:False, Scroll Lock:False, Insert:False, Win:False}
            Desactivar teclas        : {Caps Lock:False, Scroll Lock:False, Insert:False, Win:False}
        12: TOUCHPAD RAW XY        {6100} V0     
        13: GESTURE 2              {6501} V0     
            Gestos (saved): {1: True, 2: True, 10: True, 30: True, 34: False, 42: True, 43: True, 45: False, 64: False, 65: False, 67: False, 84: True}
            Gestos        : {1: True, 2: True, 30: True, 10: True, 45: False, 42: True, 43: True, 64: False, 65: False, 67: False, 84: True, 34: False}
            Gestures Diversion        : {1: False, 2: False, 10: False, 44: False, 64: False, 65: False, 67: False, 84: False, 85: False, 100: False}
            Parámetros de gestos (saved): {4: {'scale': 256}}
            Parámetros de gestos        : {4: {'scale': 256}}
        14: DFUCONTROL UNSIGNED    {00C1} V0     
        15: unknown:1811           {1811} V0    internal, hidden 
        16: unknown:1830           {1830} V0    internal, hidden 
        17: unknown:1890           {1890} V0    internal, hidden 
        18: unknown:1DF3           {1DF3} V0    internal, hidden 
        19: unknown:1E00           {1E00} V0    hidden 
        20: unknown:1EB0           {1EB0} V0    internal, hidden 
        21: unknown:1861           {1861} V0    internal, hidden 
        22: unknown:18B0           {18B0} V0    internal, hidden 
     Has 20 reprogrammable keys:
         0: Second Left Click         , default: Left Click                  => Left Click                
             mse, divertable, pos:0, group:0, group mask:empty
             reporting: default
         1: Back                      , default: Mouse Back Button           => Mouse Back Button         
             is FN, FN sensitive, reprogrammable, divertable, pos:1, group:0, group mask:empty
             reporting: default
         2: My Home                   , default: HomePage                    => HomePage                  
             is FN, FN sensitive, reprogrammable, divertable, pos:2, group:0, group mask:empty
             reporting: default
         3: Application Switcher      , default: Application Switcher        => Application Switcher      
             is FN, FN sensitive, reprogrammable, divertable, pos:3, group:0, group mask:empty
             reporting: default
         4: Contextual Menu           , default: Right Click                 => Right Click               
             is FN, FN sensitive, reprogrammable, divertable, pos:4, group:0, group mask:empty
             reporting: default
         5: Search                    , default: Search Files                => Search Files              
             is FN, FN sensitive, reprogrammable, divertable, pos:5, group:0, group mask:empty
             reporting: default
         6: Show Desktop              , default: Show Desktop                => Show Desktop              
             is FN, FN sensitive, reprogrammable, divertable, pos:6, group:0, group mask:empty
             reporting: default
         7: Maximize Window           , default: Maximize Window             => Maximize Window           
             is FN, FN sensitive, reprogrammable, divertable, pos:7, group:0, group mask:empty
             reporting: default
         8: Switch Screen             , default: Switch Presentation/Switch Screen => Switch Presentation/Switch Screen
             is FN, FN sensitive, reprogrammable, divertable, pos:8, group:0, group mask:empty
             reporting: default
         9: Media Player              , default: Music                       => Music                     
             is FN, FN sensitive, reprogrammable, divertable, pos:9, group:0, group mask:empty
             reporting: default
        10: Previous                  , default: Previous                    => Previous                  
             is FN, FN sensitive, divertable, pos:10, group:0, group mask:empty
             reporting: default
        11: Play/Pause                , default: Play/Pause                  => Play/Pause                
             is FN, FN sensitive, divertable, pos:11, group:0, group mask:empty
             reporting: default
        12: Next                      , default: Next                        => Next                      
             is FN, FN sensitive, divertable, pos:12, group:0, group mask:empty
             reporting: default
        13: Fn Left Click             , default: ShowUI                      => ShowUI                    
             divertable, pos:0, group:0, group mask:empty
             reporting: default
        14: Left Button               , default: Left Click                  => Left Click                
             mse, pos:0, group:0, group mask:empty
             reporting: default
        15: Right Button              , default: Right Click                 => Right Click               
             mse, pos:0, group:0, group mask:empty
             reporting: default
        16: Mute                      , default: Mute                        => Mute                      
             nonstandard, divertable, pos:0, group:0, group mask:empty
             reporting: default
        17: Volume Up                 , default: Volume Up                   => Volume Up                 
             nonstandard, divertable, pos:0, group:0, group mask:empty
             reporting: default
        18: Volume Down               , default: Volume Down                 => Volume Down               
             nonstandard, divertable, pos:0, group:0, group mask:empty
             reporting: default
        19: Sleep                     , default: Sleep                       => Sleep                     
             reprogrammable, divertable, pos:0, group:0, group mask:empty
             reporting: default
     Has 17 gesture(s), 1 param(s) and 5 spec(s):
        Track1Finger               Enabled(None): None   Diverted:(None) None
        TapGestures                Enabled(None): None   Diverted:(None) None
        Tap1Finger                 Enabled(   0): True   Diverted:(   0) False
        Tap2Finger                 Enabled(   1): True   Diverted:(   1) False
        TapDrag1Finger             Enabled(   2): True   Diverted:(None) None
        DoubleTap1Finger           Enabled(   3): True   Diverted:(   2) False
        Scroll2FingerStateless     Enabled(None): None   Diverted:(   3) False
        NaturalScrolling           Enabled(   4): False  Diverted:(None) None
        Scroll2FingerHoriz         Enabled(   5): True   Diverted:(None) None
        Scroll2FingerVert          Enabled(   6): True   Diverted:(None) None
        LeftEdgeSwipe1Finger2      Enabled(   7): False  Diverted:(   4) False
        RightEdgeSwipe1Finger2     Enabled(   8): False  Diverted:(   5) False
        TopEdgeSwipe1Finger2       Enabled(   9): False  Diverted:(   6) False
        Zoom2FingerStateless       Enabled(  10): True   Diverted:(   7) False
        FnClickGestureSuppression  Enabled(  11): False  Diverted:(None) None
        TwoFingersPresent          Enabled(None): None   Diverted:(   8) False
        DeviceSpecificRawData      Enabled(None): None   Diverted:(   9) False
        ScaleFactor                Value  (   0): 256 [Default: 256]
        period unit                Spec   (   3): 2048
        DVI field width            Spec   (   1): 8
        field widths               Spec   (   2): 8
        multiplier                 Spec   (   5): None
        resolution                 Spec   (   4): 4
     Battery: 90%, discharging, next level 50%.

```
</details>

- Contents of `~/.config/solaar/config.yaml` (or `~/.config/solaar/config.json` if `~/.config/solaar/config.yaml` not present):

<details>

```
CONTENTS HERE
```
</details>

- Errors or warrnings from Solaar:

**Describe the bug**
simply put, that it doesn't fn-swap when the program is run automatically by the window manager

**To Reproduce**
try to add it to xfce session startup apps or xinitrc, it runs but doesn't fn swap (can't do ALT+F4 without geting the stupid dialog menu instead of closing a window)

**Additional context**
just running mint with XFCE, nothing too fancy

**Comment by pfps** [comment]
Diagnosing problems during startup is a pain.  It appears that something is preventing `solaar config` from doing the right thing.    One thing to check is whether the Solaar GUI is also running at startup,  if so there might be problems due to interactions between the GUI and the CLI.  If you do run the Solaar GUI then there is no need to use `solaar config` as the GUI will remember the setting for fn-swap and push it to the device.

If you are not running the Solaar GUI then my suggestion is to write a simple script that captures all the output from `solaar config` so that you can see what is happening.   Something like
```
#! /bin/sh

solaar config "K400 Plus" fn-swap > /tmp/solaar 2>&1
solaar -ddd config "K400 Plus" fn-swap >> /tmp/solaar 2>&1
solaar config "K400 Plus" fn-swap >> /tmp/solaar 2>&1
```

If you can't determine what is going wrong from that output, please post it here.

**Comment by webodan** [comment]
Thanks for the fast response!

This is what it gave me after running it right now, couldn't understand it. Something to note: I had manually swapped the fn keys prior to this

`dan@htpc:~$ cat /tmp/solaar 
# Intercambiar función Fx
# Al activarse, las teclas F1..F12 activarán sus funciones especiales, y debe mantener pulsada la tecla FN para activar su función estándar.  Al desactivarse, las teclas F1..F12 activarán sus funciones estándar, y debe mantener pulsada la tecla FN para activar su función especial.
#   possible values: on/true/t/yes/y/1 or off/false/f/no/n/0 or Toggle/~
fn-swap = False
2022-10-05 12:28:35,678,678    DEBUG [MainThread] hidapi.udev: Found device BID 0003 VID 0000046D PID 0000C52B INTERFACE 2 FILTER 2
2022-10-05 12:28:35,679,679     INFO [MainThread] logitech_receiver.base: New lock 4
2022-10-05 12:28:35,679,679    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 83B5 030000]
2022-10-05 12:28:35,681,681    DEBUG [MainThread] logitech_receiver.base: (4) => r[11 FF 83B5 0342F4925D0106070000000000000000]
2022-10-05 12:28:35,682,682    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 80B2 000000]
2022-10-05 12:28:35,687,687    DEBUG [MainThread] logitech_receiver.base: (4) => r[20 01 4101 4D401E4000000400000000]
2022-10-05 12:28:35,687,687    DEBUG [MainThread] logitech_receiver.base: (4) => r[20 00 4102 0000000000000000000000]
2022-10-05 12:28:35,689,689    DEBUG [MainThread] logitech_receiver.base: (4) => r[10 FF 8F80 B20300]
2022-10-05 12:28:35,689,689    DEBUG [MainThread] logitech_receiver.base: (4) device 0xFF error on request {80B2}: 3 = invalid value
2022-10-05 12:28:35,690,690    DEBUG [MainThread] solaar.cli: [/dev/hidraw8] => <ReceptorUnifying(/dev/hidraw8,4)>
2022-10-05 12:28:35,738,738    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 8102 000000]
2022-10-05 12:28:35,741,741    DEBUG [MainThread] logitech_receiver.base: (4) => r[10 FF 8102 000100]
2022-10-05 12:28:35,743,743    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 83B5 200000]
2022-10-05 12:28:35,746,746    DEBUG [MainThread] logitech_receiver.base: (4) => r[11 FF 83B5 200708404D0402014700000000000000]
2022-10-05 12:28:35,746,746    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 83B5 300000]
2022-10-05 12:28:35,749,749    DEBUG [MainThread] logitech_receiver.base: (4) => r[11 FF 83B5 30304E66B31E40000009000000000000]
2022-10-05 12:28:35,749,749     INFO [MainThread] logitech_receiver.receiver: <ReceptorUnifying(/dev/hidraw8,4)>: found new device 1 (404D)
2022-10-05 12:28:35,749,749    DEBUG [MainThread] logitech_receiver.base: (12) pinging device 1
2022-10-05 12:28:35,749,749     INFO [MainThread] logitech_receiver.base: New lock 12
2022-10-05 12:28:35,750,750    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 001D 00005C00000000000000000000000000]
2022-10-05 12:28:35,835,835    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 001D 04015C00000000000000000000000000]
2022-10-05 12:28:35,836,836    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 000F 00010000000000000000000000000000]
2022-10-05 12:28:35,857,857    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 000F 01000100000000000000000000000000]
2022-10-05 12:28:35,857,857    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 010A 00000000000000000000000000000000]
2022-10-05 12:28:35,877,877    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 010A 17000000000000000000000000000000]
2022-10-05 12:28:35,878,878    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 000F 40A00000000000000000000000000000]
2022-10-05 12:28:35,897,897    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 000F 00000000000000000000000000000000]
2022-10-05 12:28:35,897,897    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 000C 40A20000000000000000000000000000]
2022-10-05 12:28:35,917,917    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 000C 09000000000000000000000000000000]
2022-10-05 12:28:35,917,917    DEBUG [MainThread] logitech_receiver.settings_templates: check_feature fn-swap [NEW FN INVERSION] detected <Setting([característica:toggle] K400 Plus:fn-swap=None)>
2022-10-05 12:28:35,917,917    DEBUG [MainThread] logitech_receiver.settings: fn-swap: settings read None from <Device(1,404D,Wireless Touch Keyboard K400 Plus,304E66B3)>
2022-10-05 12:28:35,923,923    DEBUG [MainThread] solaar.configuration: load => ['1.1.5', {'_NAME': 'Wireless Touch Keyboard K400 Plus', '_modelId': '404D00000000', '_sensitive': {'fn-swap': False}, '_serial': '304E66B3', '_unitId': '0F18F9DB', '_wpid': '404D', 'disable-keyboard-keys': {1: False, 4: False, 8: False, 16: False}, 'divert-keys': {1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 8: 0, 34: 0, 40: 0, 62: 0, 64: 0, 79: 0, 84: 0, 110: 0, 145: 0, 149: 0, 183: 0, 184: 0}, 'fn-swap': False, 'gesture2-gestures': {1: True, 2: True, 10: True, 30: True, 34: False, 42: True, 43: True, 45: False, 64: False, 65: False, 67: False, 84: True}, 'gesture2-params': {4: {'scale': 256}}}]
2022-10-05 12:28:35,923,923    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 000C 00030000000000000000000000000000]
2022-10-05 12:28:36,017,017    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 000C 02000100000000000000000000000000]
2022-10-05 12:28:36,017,017    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 020B 00000000000000000000000000000000]
2022-10-05 12:28:36,037,037    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 020B 030F18F9DB0004404D00000000000000]
2022-10-05 12:28:36,038,038     INFO [MainThread] logitech_receiver.device: <Device(1,404D,Wireless Touch Keyboard K400 Plus,304E66B3)>: unitId 0F18F9DB does not match serial 304E66B3
2022-10-05 12:28:36,038,038    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 090E 00000000000000000000000000000000]
2022-10-05 12:28:36,057,057    DEBUG [MainThread] logitech_receiver.base: (12) => r[11 01 090E 00010000000000000000000000000000]
2022-10-05 12:28:36,057,057    DEBUG [MainThread] logitech_receiver.settings: BooleanValidator: validate read b'\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' => 00
# Intercambiar función Fx
# Al activarse, las teclas F1..F12 activarán sus funciones especiales, y debe mantener pulsada la tecla FN para activar su función estándar.  Al desactivarse, las teclas F1..F12 activarán sus funciones estándar, y debe mantener pulsada la tecla FN para activar su función especial.
#   possible values: on/true/t/yes/y/1 or off/false/f/no/n/0 or Toggle/~
fn-swap = False
# Intercambiar función Fx
# Al activarse, las teclas F1..F12 activarán sus funciones especiales, y debe mantener pulsada la tecla FN para activar su función estándar.  Al desactivarse, las teclas F1..F12 activarán sus funciones estándar, y debe mantener pulsada la tecla FN para activar su función especial.
#   possible values: on/true/t/yes/y/1 or off/false/f/no/n/0 or Toggle/~
fn-swap = False
`

**Comment by pfps** [comment]
Oops I made a mistake in the script, it should be 

```
#! /bin/sh

solaar config "K400 Plus" fn-swap > /tmp/solaar 2>&1
solaar -ddd config "K400 Plus" fn-swap false >> /tmp/solaar 2>&1
solaar config "K400 Plus" fn-swap >> /tmp/solaar 2>&1
```

You should run this script at startup and see what output is produced.

**Comment by webodan** [comment]
@pfps in order to not clutter the thread, i've updated the output with what i got in the previous comment, looks like there's a bug

**Comment by pfps** [comment]
The output indicates that the keyboard is disconnected.  This is a problem running Solaar commands at startup or even user login - the desired device might not be connected.   The Solaar GUI is better, as it delays sending commands to devices until they are connected.

Another problem with using `solaar config` to change the behaviour of devices is that devices forget most settings when they go into a power saving state.   The Solaar GUI notices when devices wake up and sends commands to them to restore their settings when they do.

**Comment by webodan** [comment]
all i can say about it is... the keyboard's always *physically* connected to that machine... since it's an htpc and i use it on a TV and it's my only means of using the computer. if something is causing the issue, it's definitely something else which i don't know about

**Comment by pfps** [comment]
Even if the receiver was connected and active the keyboard was in a power saving mode, as evidenced by Solaar's attempt to connect to the keyboard.  

```
2022-10-05 14:36:32,746,746    DEBUG [MainThread] logitech_receiver.base: (12) pinging device 1
2022-10-05 14:36:32,746,746     INFO [MainThread] logitech_receiver.base: New lock 12
2022-10-05 14:36:32,746,746    DEBUG [MainThread] logitech_receiver.base: (12) <= w[11 01 0018 00008C00000000000000000000000000]
2022-10-05 14:36:32,750,750    DEBUG [MainThread] logitech_receiver.base: (12) => r[10 01 8F00 180900]
```

The last debug message shows an error response from the receiver when trying to ping the keyboard.  The error code in the response, 0x09, has the following description in Logitech documents:

5.6 ERR_RESOURCE_ERROR (0x09)
This error is returned by the receiver when an hidpp command has been sent to a device that is in disconnected mode. When a device is in disconnected mode it cannot receive commands from the host until it reconnects. A device reconnects when the user interacts with it. In most cases, a device disconnects after several minutes of inactivity.

**Comment by webodan** [comment]
Any way i could work around that? like, maybe software-reconnecting the device then running the command so it will apply the fn-swap correctly?

**Comment by pfps** [comment]
The device is disconnected and will only be connected when something happens on the device so there is no way to have software cause the device to connect.  The Solaar GUI gets around this problem by responding to the message that the device sends when it connects.   

There is still a problem in the Solaar GUI - pressing a key on the device when it is disconnected causes the device to connect but also does whatever is the default action for the key as Solaar has not had a chance to send commands to the device to modify its behaviour.   There is a way around this but it has not (yet) been implemented in Solaar.  For now, the workaround is to run the Solaar GUI and wake up devices with a neutral action.  For the K400 this would be something like moving the mouse pointer.

**Comment by pfps** [comment]
Closing as the solution is to use the GUI.
